core/shared/src/main/scala/zio/logging/internal/WriterProvider.scala

```diff
@@ -0,0 +1,81 @@
+package zio.logging.internal
+
+import java.io.{ BufferedWriter, FileOutputStream, OutputStreamWriter, Writer }
+import java.nio.charset.Charset
+import java.nio.file.{ FileSystems, Path }
+import java.time.LocalDateTime
+import java.time.format.DateTimeFormatter
+import java.time.temporal.ChronoUnit
+
+private[logging] sealed trait WriterProvider {
+  def writer: Writer
+}
+
+private[logging] object WriterProvider {
+  final case class SimpleWriterProvider(
+    destination: Path,
+    charset: Charset,
+    bufferedIOSize: Option[Int]
+  ) extends WriterProvider {
+    override val writer: Writer = {
+      val output = new OutputStreamWriter(new FileOutputStream(destination.toFile, true), charset)
+      bufferedIOSize match {
+        case Some(bufferSize) => new BufferedWriter(output, bufferSize)
+        case None             => output
+      }
+    }
+  }
+
+  final case class TimeBasedRollingWriterProvider(
+    destination: Path,
+    charset: Charset,
+    bufferedIOSize: Option[Int]
```

Comment:
it will be good if we can test it

if we add here another attribute

```scala 
makeNewTime: () => LocalDateTime = TimeBasedRollingWriterProvider.makeNewTime
```

then in WriterProviderSpec

it should be possible to do something like:

```scala
  val timeRef = new AtomicReference[LocalDateTime](LocalDateTime.now())
  
   timeRef.set(LocalDateTime.now().plusDays(1))
  
   val testMakeNewTime: () => LocalDateTime = () => timeRef.get()
```
and then test it if in case of parallel execution, when  `WriterProvider.writer` is called multiple times with same time, if it return same writer

Reply:
I wrote test code. Could you check this commit? https://github.com/zio/zio-logging/pull/691/commits/4fe78ca9e9229d06e767afb044df6ae36fe4b434